sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -58,13 +57,28 @@ abstract class QueryPlan[PlanType <: QueryPlan[PlanType]] extends TreeNode[PlanT
   }
 
   /**
+   * Infer the Attribute-specific IsNotNull constraints from the null intolerant child expressions
+   * of constraints.
+   */
+  private def inferIsNotNullConstraints(constraint: Expression): Seq[Expression] =
```

Comment:
This change simply ignores all `IsNotNull`s which are not the top expression. The above case works because `Filter` splits it. But if the constraint looks like `Cast(IsNotNull(a), Integer) == 1`, we won't infer `IsNotNull(a)` from it, right?